README.md
# NBA Ticker 🏀

![Visual Studio Marketplace Version](https://img.shields.io/visual-studio-marketplace/v/omkarmoghe.nba-ticker?label=Release)

Adds live NBA scores for today's games in your status bar using the data.nba.net API.

![Status bar ticker](https://i.imgur.com/EfQzKgb.gif)

![Game details](https://i.imgur.com/dW8HKoh.png)
![Scoreboard](https://i.imgur.com/QAJuS9T.png)

## Requirements

- VSCode must have a working internet connection.

## Installation

Install via the [VSCode Marketplace](https://marketplace.visualstudio.com/items?itemName=omkarmoghe.nba-ticker).

## Settings

All settings are prefixed by with `nba-ticker.`, so `side` becomes `"nba-ticker.side"` when editing as JSON.

### `format`
Specifies how the game info should be displayed. Defaults to `"${vTeam} ${vScore} - ${hScore} ${hTeam}, ${status}"`, which looks like `GSW 42 - 24 LAL, 4:20 1st`. The available wildcards are:
- `${vTeam}`: The visiting team's abbreviated name, (e.g. `"GSW"`)
- `${vScore}`: The visiting team's score
- `${hTeam}`: The home team's abbreviated name, (e.g. `"LAL"`)
- `${hScore}`: The home team's score
- `${status}`: The status of the game
  - If the game hasn't started yet, this will be the scheduled start time (e.g. `"9:30 PM"`) in your local UTC offset.
  - If the game is ongoing, this will be either `"<TIME LEFT IN PERIOD> <PERIOD>"` (e.g. `"4:20 1st"`), `"End of 1st"`, or `"Halftime"`
  - If the game is over, this will be `"Final"`
  - If the status is unknown, this will be `"TBD"`

### `hideFinalScores`
If checked, final scores will be hidden from the status bar. Final scores will always be visible in the status bar hover when the `hover` option is set to `scoreboard`. Defaults to `false`.

### `hover`
Specifies what to show when hovering over the ticker. Options are `"details"` (details about the current game) or `"scoreboard"` (all scores).

### `pollDelaySeconds`
Specifies how long to wait before fetching new game data from the API. Defaults to `60` seconds, i.e. every minute. Allowed values are [`10`, `600`].

### `priority`
Specifies the ticker's priority relative to other items in the status bar. **Higher values are shows more to the left.**

### `side`
Specifies which side of the status bar to display the ticker. Options are `"left"` or `"right"`.

### `teamFilter`
Specifies which teams games are shown for; other teams are ignored. All teams will be shown if the array is empty (`[]`). **Use abbreviated team names** (e.g. `["GSW", "LAL"]`).

### `tickerDelaySeconds`
Specifies how many seconds to show each score. Defaults to `10` seconds. Allowed values are [`1`, `60`].

## FAQ

### How fast is "live"?
- All versions >= `0.3.0` (including current): scores update as fast as the data.nba.net API.
- For versions < `0.3.0`: scores update every ~10 minutes.

## [Changelog](./CHANGELOG.md)

## Credits
- Versions before `0.3.0` were built using the amazing [balldontlie API](https://www.balldontlie.io/#introduction). If you like this extension, please consider [donating](https://www.patreon.com/balldontlie) to them.
- Icon by [Cahya Kurniawa](https://www.iconfinder.com/sikeystudio) via [Iconfinder](https://www.iconfinder.com/).


CHANGELOG.md
# Change Log

All notable changes to the "nba-ticker" extension will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.6.4] - 2023-12-26
### Fixed
- Fixed bug where a stale command (link) would be left on the status bar.
### Changed
- Dependency upgrades

## [0.6.3] - 2023-11-05
### Added
- `hideFinalScores` setting to skip final scores in the status bar
### Changed
- Dependency upgrades

## [0.6.2] - 2023-06-03
### Changed
- Dependency upgrades

## [0.6.1] - 2023-05-20
### Added
- Stops polling for games when all games are `Final`. This is reset when the calendar day changes.

### Changed
- Migrated from `moment` -> `day.js`

## [0.6.0] - 2023-05-19
### Added
- New `Open Game` command to open a game's page from the command palette.

### Changed
- Dependency updates

## [0.5.2] - 2022-12-03
### Changed
- Added additional check to see if the game period has actually started.

## [0.5.1] - 2022-12-03
### Changed
- Updated to latest working NBA API
- Fixed broken URLs
- Changed hover default from `details` -> `scoreboard`

## [0.4.4] - 2022-11-26
### Changed
- New name and icon!

## [0.4.3] - 2022-11-26
### Changed
- Bumped dependencies
- Updated NBA API protocol from HTTP -> HTTPS
- Fixed issue where links were broken for games that hadn't started
- Fixed issue where games that had just started would render as `0th`. This now displays the tipoff time until a valid period [1-4] and game clock is available.

## [0.4.2] - 2021-12-04
### Changed
- Bumped dependencies
- Updated default format:
  - `@` -> `, `
  - `:` -> `-`
- Updated description

## [0.4.1] - 2021-03-21
### Changed
- Updated gif in README

## [0.4.0] - 2021-03-21
### Added
- New hover view shows all games instead of game details (option in settings)
### Changed
- Bumped various dependencies

## [0.3.7] - 2020-12-17
### Changed
- Fixed an issue where empty game "nugget" text resulted in an empty line in the tooltip

## [0.3.6] - 2020-12-14
### Changed
- Bump dependencies

## [0.3.5] - 2020-11-11
### Added
- README release version shield

## [0.3.4] - 2020-10-26
### Changed
- Fixed minimum poll delay validation

## [0.3.3] - 2020-10-24
### Changed
- Bumped project Node version from `14.13.1` -> `15.0.0`

## [0.3.2] - 2020-10-17
### Added
- Clicking on the status bar Ticker opens the game in https://stats.nba.com
### Changed
- Fixed settings & README

## [0.3.1] - 2020-10-17
### Changed
- Fixed images on Marketplace page

## [0.3.0] - 2020-10-17
### Changed
- Migrated from the balldontlie API to official(?) data.nba.net API
- Updated status
### Added
- Huge update to the tooltip with a ton of game and series details

## [0.2.5] - 2020-10-17
### Changed
- Bumped Node to `14.13.1`
- Bumped Typescript & dependencies to `^4`
- Bumped ESLint, Moment, Axios

## [0.2.4] - 2020-09-08
### Changed
- Fixed typo in file/module naming

## [0.2.3] - 2020-09-08
### Changed
- Fixed typo in settings

## [0.2.2] - 2020-09-08
### Changed
- Updated default ticker display format.
- Updated ticker tooltip to show relative time of update (e.g. `"5 minutes ago"`)

## [0.2.1] - 2020-08-25
### Added
- Setting for teams
### Changed
- Updated default format separator from `|` to `-`

## [0.2.0] - 2020-08-25
### Added
- Setting for poll frequency
- Settings for status bar placement, format, and duration
### Changed
- Breaking changes to registered commands

## [0.0.2] - 2020-08-19
### Changed
- Fixed date to use current date

## [0.0.1] - 2020-08-19
### Added
- Status bar item to display scores
- Command to update scores (no UI)


src/balldontlie-api.ts
import axios from "axios";
import * as moment from "moment";

const API_VERSION: Readonly<string> = "v1";
const BASE_URL: Readonly<string> = `https://balldontlie.io/api/${API_VERSION}`;
const GAMES: Readonly<string> = `${BASE_URL}/games`;

const teamMap: Readonly<Record<string, number>> = {
  ATL: 1,
  BOS: 2,
  BKN: 3,
  CHA: 4,
  CHI: 5,
  CLE: 6,
  DAL: 7,
  DEN: 8,
  DET: 9,
  GSW: 10,
  HOU: 11,
  IND: 12,
  LAC: 13,
  LAL: 14,
  MEM: 15,
  MIA: 16,
  MIL: 17,
  MIN: 18,
  NOP: 19,
  NYK: 20,
  OKC: 21,
  ORL: 22,
  PHI: 23,
  PHX: 24,
  POR: 25,
  SAC: 26,
  SAS: 27,
  TOR: 28,
  UTA: 29,
  WAS: 30,
};

interface Team {
  abbreviation: string,
  city: string,
  conference: string,
  division: string,
  full_name: string,
  id: number,
  name: string,
}

export interface Game {
  id: number,
  period: number,
  date: string,
  season: number,
  postseason: boolean,
  status: string,
  time: string,
  home_team: Team,
  home_team_score: number,
  visitor_team: Team,
  visitor_team_score: number,
}

export class Score {
  visitorTeam: string;
  visitorScore: number;
  homeTeam: string;
  homeScore: number;
  status: string;

  constructor(game: Game) {
    this.visitorTeam = game.visitor_team.abbreviation.trim();
    this.visitorScore = game.visitor_team_score;
    this.homeTeam = game.home_team.abbreviation.trim();
    this.homeScore = game.home_team_score;
    // Determine status string.
    if (game.time.trim()) {
      this.status = `${game.time.trim()} ${this.humanPeriod(game.period)}`;
    } else {
      this.status = game.status.trim() || "TBD";
    }
  }

  humanPeriod(period: number): string {
    switch (period) {
      case 1:
        return `${period}st`;
      case 2:
        return `${period}nd`;
      case 3:
        return `${period}rd`;
      case 4:
      default:
        return `${period}th`;
    }
  }

  format(formatString: string) {
    return formatString
      .replace(/\${visitorTeam}/gi, this.visitorTeam)
      .replace(/\${visitorScore}/gi, this.visitorScore.toString())
      .replace(/\${homeTeam}/gi, this.homeTeam)
      .replace(/\${homeScore}/gi, this.homeScore.toString())
      .replace(/\${status}/gi, this.status);
  }
}

const today = () => moment().format("YYYY-MM-DD");

export function fetchGames(teams: string[] = []): Promise<Game[]> {
  const params: Record<string, string[] | number[]> = {
    dates: [today()]
  };

  if (teams.length) {
    params.team_ids = teams.map((teamAbbrev) => teamMap[teamAbbrev]);
  }

  return axios.get(GAMES, { params }).then(({ status, statusText, data }) => {
    console.info(`GET ${GAMES} ${status} ${statusText}`);
    return data.data;
  });
}
